crates/tsk/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+            {
+                use std::os::windows::ffi::OsStrExt as _;
+                path.as_os_str()
+                    .encode_wide()
+                    .chain(Some(0))
+                    .collect::<Vec<u16>>()
+            }
+        };
+        let tsk_img_result = unsafe {
+            tsk_sys::tsk_img_open_sing(
+                path.as_ptr(),
+                tsk_sys::TSK_IMG_TYPE_ENUM_TSK_IMG_TYPE_RAW_SING,
+                0,
+            )
+        };
+        get_tsk_result(tsk_img_result).map(|inner| Self { inner })
+    }
+}
+
+impl Drop for TskImage {
+    fn drop(&mut self) {
+        unsafe { tsk_sys::tsk_img_close(self.inner.as_mut()) };
+    }
+}
+
+pub struct TskFs<'a> {
+    pub(crate) inner: NonNull<tsk_sys::TSK_FS_INFO>,
+    _marker: PhantomData<&'a tsk_sys::TSK_FS_INFO>,
+}
+
+pub enum WalkDirCallbackResult {
+    /// Continue walking.
+    Continue,
+    /// Stop walking.
+    Stop,
+}
+
+impl TskFs<'_> {
+    pub fn open(image: &TskImage) -> TskResult<Self> {
+        let tsk_fs_result = unsafe {
+            tsk_sys::tsk_fs_open_img(image.inner.as_ptr(), 0, TSK_FS_TYPE_ENUM_TSK_FS_TYPE_DETECT)
+        };
+        get_tsk_result(tsk_fs_result).map(|inner| TskFs {
+            inner,
+            _marker: PhantomData,
+        })
+    }
+
+    pub fn root_inum(&self) -> u64 {
+        unsafe { self.inner.as_ref() }.root_inum
+    }
+    /// Returns the string name of a file system type id, e.g. "ntfs".
+    pub fn get_fs_type(&self) -> TskResult<String> {
+        let ty = unsafe { self.inner.as_ref().ftype };
+        let name_ptr = unsafe { tsk_sys::tsk_fs_type_toname(ty) };
+        get_tsk_result(name_ptr as _)
+            .map(|non_null| unsafe { CStr::from_ptr(non_null.as_ptr()).to_bytes() })
+            .map(|bytes| String::from_utf8_lossy(bytes).to_string())
+    }
+
+    /// Opens a directory given its path.
+    pub fn open_dir(&self, path: &Path) -> TskResult<TskFsDir> {
+        let tsk_path = TskPath::from_path(path);
+        let result = unsafe { tsk_fs_dir_open(self.inner.as_ptr(), tsk_path.as_ptr()) };
+        get_tsk_result(result as _).map(TskFsDir::new)
+    }
+
+    /// Opens a file given its metadata address.
+    pub fn open_dir_meta(&self, meta: u64) -> TskResult<TskFsDir> {
+        let result = unsafe { tsk_fs_dir_open_meta(self.inner.as_ptr(), meta) };
+        get_tsk_result(result as _).map(TskFsDir::new)
+    }
+
+    /// Opens a file given its path.
+    pub fn open_file(&self, path: &Path) -> TskResult<TskFsFile> {
+        let tsk_path = TskPath::from_path(path);
+        let result =
+            unsafe { tsk_fs_file_open(self.inner.as_ptr(), null_mut(), tsk_path.as_ptr()) };
+        get_tsk_result(result as _).map(TskFsFile::new)
+    }
+
+    /// Recursively walks the directory with the given metadata address. Calls
+    /// `callback` for each file and directory.
+    pub fn walk_dir(
+        &mut self,
+        dir_addr: u64,
```

Comment:
Given how we are using this method, why not make it slightly more type safe and have it accept `TskFsDir`?